### ISSUE TITLE
some general questions

### ISSUE THREAD

**Issue by fabisch30** [vasturiano/3d-force-graph#282]
Hi,

i am still using this library with pleasure. I come up with two (maybe more general) questions but they came up during my work with this, so i asked them here:

1) I have an button where i can select an json file (with the correct format) which then will be immediately displayed. I use for that:

`<input type="file" id="myFile" accept=".json">`

Problem is if i select an json file with an error (eg. source id of link does not exist) or not at all a json file and then select a correct json file, it still tries to load the first incorrect file. Is there a way to stop the loading of the first incorrect file?

I simply use `Graph.graphData(JSON.parse(jsonFile))` for loading.

2) You have an example where you can change visualisation features dynamically with sliders ([example](https://vasturiano.github.io/3d-force-graph/example/manipulate-link-force/). I want to do exactly the same, but i want my sliders in an <div> element which i can place wherever i want (also want to hide it and expand if you click an that button; this already implemented with other features). Question is how do i place my sliders in an html <div> element. Because right now you have the sliders (buttons) defined in the javascript part. I want them in my html part like this:

<div id="Sliders" >
    <button> slider1 </button>
    <button> slider2  </button>
</div>

Best regards,
Fabian.

**Comment by Alexithemia** [comment]
1. You would need to pre-process the data before sending it to the graph library.

```
const processData = (data) => {
  //check data has keys - nodes & links, and that they are both arrays. If not return null
  //data.nodes.filter - only return nodes that have an id (that is all that is required by the library)
  //data.links.filter - only return link where both source and target exists in data.nodes or add a placeholder node to data.nodes for it.
}

const jsonData = JSON.parse(jsonFile);

Graph.graphData(processData(jsonData));
```

2. You would need to add a click or on change event handler to the HTML to fire a function in the javascript.

**Comment by fabisch30** [comment]
Thanks a lot for answering.

1) I was hoping there would be a function you can call which stops the loading of graphData. Seems there isn`t one. I get your idea, but really don`t know how i would implement that. I will look this up...

2) But don`t I need to specifiy where the sliders are placed. Of course I could put all the javascript for the Slider in a function which is called on an clickEvent, but don`t I have the same problems then? The Slider are defined in the javascript part. That`s the problem here.

**Comment by fabisch30** [comment]
Also there are more criteria then just _node has id_ and _link source/target exists_. For example if the Name of an node consists of quotes (because they are badly extracted from an database), then also the graph can not be displayed.
So a  function which just stops the upload/rendering of the graph data would be really helpful. Also sometimes the graph is too big and takes to long to display I want to stop the rendering of it and load another one. 
How would I do this then?

**Comment by Alexithemia** [comment]
If you don't call `Graph.graphData()` with a new set of data it will not try to load it. You will need to import the jsonfile and check it yourself to make sure the data is formatted correctly. If it is then you would call `Graph.graphData(jsonfile)`, if something is wrong with the data, just don't call `Graph.graphData()` on the data and instead throw a warning or error for the user to see.

There's no direct way to stop the loading of graph I could think of, you could probably just call `Graph.graphData()` again with no data to empty it, but I am not sure that would stop the process it is going through. You could try that, if that doesn't work you can try calling `Graph.cooldownTime(0)` to stop the force engine from moving the nodes around which is likely using most of the processing power.

**Comment by fabisch30** [comment]
> you could probably just call Graph.graphData() again with no data to empty it

 does unfortunately not work.

> You will need to import the jsonfile and check it yourself to make sure the data is formatted correctly.

 there are lot of possible things that could go wrong. so a lot to implement, because as I said its not just **node has id** and **link source/target exists** to test.

> You could try that, if that doesn't work you can try calling Graph.cooldownTime(0) to stop the force engine from moving the nodes around which is likely using most of the processing power. l

ets try that....

**Comment by Alexithemia** [comment]
Anything other than Node having ID and links source and target being a string of a valid Node ID, is something you may have added to customize the graph and you would of course need to check that. I'm just saying those are the minimum this library needs to render a node or link.

You could also try ` Graph.pauseAnimation()` but still unsure what will work depending on where the graph is stuck loading at.

There is no integrated way to stop the graph process, so If the graph is in the middle of loading a large amount of data and is taking awhile, the only surefire thing to do is to remove the component the graph is attached to, set the Graph variable used to null, and hope the garbage collection cleans it up sooner rather than later. Currently trying to solve an issue related to garbage collection and performance myself.

**Comment by fabisch30** [comment]
> remove the component the graph is attached to

what do you mean with that? 

> set the Graph variable used to null

if i do GraphData(data) where data is an empty graph (no nodes/links) i still have the problem i had first that is the machine still trys to load the first graph.

**Comment by Alexithemia** [comment]
Deleting the HTML element the graph is attached to.

And yeah depending on where the process is in trying to graph the data there could be no way to stop it. Maybe look into cutting the graph data into chunks and loading them in one at a time? With checks to break the loading process inbetween? Maybe add the nodes, then add the links?

The library can handle 4,000 objects pretty well if you check his example: https://vasturiano.github.io/3d-force-graph/example/large-graph/

**Comment by fabisch30** [comment]
Yeah so if the graph is too big and you want to stop the rendering, removing the html div element does not help. 

What I will do now is to set up a function (as you suggested) to pre-process/check the data.
Also will think about adding filtering options in advance for big data to shrink the size of the graph.

Will leave the issue open (and change the title) if at some time there is a solid way to stop a graph from loading/rendering. Thanks for your help!
